janus_server/src/aggregator.rs

```diff
@@ -183,38 +223,299 @@ impl<C: Clock> Aggregator<C> {
             .await?;
         Ok(())
     }
+
+    /// Implements the `/aggregate` endpoint for the helper, described in §4.4.4.1 & §4.4.4.2 of
+    /// draft-gpew-priv-ppm.
+    async fn handle_aggregate(&self, req: AggregateReq) -> Result<AggregateResp, Error> {
+        match req.body {
+            AggregateInitReq { agg_param, seq } => {
+                self.handle_aggregate_init(req.task_id, req.job_id, agg_param, seq)
+                    .await
+            }
+            AggregateContinueReq { seq } => {
+                self.handle_aggregate_continue(req.task_id, req.job_id, seq)
+                    .await
+            }
+        }
+    }
+
+    /// Implements the aggregate initialization request portion of the `/aggregate` endpoint for the
+    /// helper, described in §4.4.4.1 of draft-gpew-priv-ppm.
+    async fn handle_aggregate_init(
+        &self,
+        task_id: TaskId,
+        job_id: AggregationJobId,
+        agg_param: Vec<u8>,
+        report_shares: Vec<ReportShare>,
+    ) -> Result<AggregateResp, Error> {
+        // If two ReportShare messages have the same nonce, then the helper MUST abort with
+        // error "unrecognizedMessage". (§4.4.4.1)
+        let mut seen_nonces = HashSet::with_capacity(report_shares.len());
+        for share in &report_shares {
+            if !seen_nonces.insert(share.nonce) {
+                return Err(Error::UnrecognizedMessage(
+                    "aggregate request contains duplicate nonce",
+                ));
+            }
+        }
+
+        // Decrypt shares & prepare initialization states. (§4.4.4.1)
+        // TODO(brandon): reject reports that are "too old" with `report-dropped`.
+        // TODO(brandon): reject reports in batches that have completed an aggregate-share request with `batch-collected`.
+        struct ReportShareData<A: vdaf::Aggregator>
```

Comment:
I disagree, this data-type is specific to this method & I don't think there's much reason to widen the visibility -- the upside is basically nil, the downside is that the code is slightly harder to understand because readers have to realize that `ReportShareData` is only used here. [If more than one method wanted to use this type, then yeah, I'd support moving it out -- but this is so situationally-specific that I don't know we'll need exactly the same type.]

I minimalized the trait bounds, but unfortunately I still need one bound.